src/Microsoft.ML.TorchSharp/Utils/Range.cs

```diff
@@ -134,5 +137,293 @@ public override string ToString()
+        }
+
+        /// <summary>
+        /// Returns a reference to the 0th element of <paramref name="array"/>. If the array is empty, returns a reference to where the 0th element
+        /// would have been stored. Such a reference may be used for pinning but must never be dereferenced.
+        /// </summary>
+        /// <exception cref="NullReferenceException"><paramref name="array"/> is <see langword="null"/>.</exception>
+        /// <remarks>
+        /// This method does not perform array variance checks. The caller must manually perform any array variance checks
+        /// if the caller wishes to write to the returned reference.
+        /// </remarks>
+        [Intrinsic]
+        public static ref T GetArrayDataReference<T>(T[] array) =>
+            ref array[0];
+
+        /// <summary>
+        /// Returns a reference to the 0th element of <paramref name="array"/>. If the array is empty, returns a reference to where the 0th element
+        /// would have been stored. Such a reference may be used for pinning but must never be dereferenced.
+        /// </summary>
+        /// <exception cref="NullReferenceException"><paramref name="array"/> is <see langword="null"/>.</exception>
+        /// <remarks>
+        /// The caller must manually reinterpret the returned <em>ref byte</em> as a ref to the array's underlying elemental type,
+        /// perhaps utilizing an API such as <em>System.Runtime.CompilerServices.Unsafe.As</em> to assist with the reinterpretation.
+        /// This technique does not perform array variance checks. The caller must manually perform any array variance checks
+        /// if the caller wishes to write to the returned reference.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe ref byte GetArrayDataReference(Array array)
+        {
+            // If needed, we can save one or two instructions per call by marking this method as intrinsic and asking the JIT
+            // to special-case arrays of known type and dimension.
+
+            // See comment on RawArrayData (in RuntimeHelpers.CoreCLR.cs) for details
+            return ref Unsafe.AddByteOffset(ref Unsafe.As<RawData>(array).Data, (nuint)GetMethodTable(array)->BaseSize - (nuint)(2 * sizeof(IntPtr)));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe MethodTable* GetMethodTable(object obj)
+        {
+            // The body of this function will be replaced by the EE with unsafe code
+            // See getILIntrinsicImplementationForRuntimeHelpers for how this happens.
+
+            return (MethodTable*)Unsafe.Add(ref Unsafe.As<byte, IntPtr>(ref GetRawData(obj)), -1);
+        }
+
+        internal static ref byte GetRawData(object obj) =>
+            ref Unsafe.As<RawData>(obj).Data;
+
+        internal sealed class RawData
+        {
+            public byte Data;
+        }
+
+#if TARGET_ARM64 || TARGET_LOONGARCH64
```

Comment:
These IFDEF's aren't going to work for us since we don't cross-compile our managed library, you might want to refactor as property initialized based on `RuntimeInformation.ProcessorArchitecture`.